File: python/perfect-numbers/perfect_numbers.py
"""Perfect numbers."""


def classify(number):
    """Classify ``number`` as perfect, abundant or deficient."""

    if number <= 0:
        raise ValueError("Classification is only possible for positive integers.")
    return ["perfect", "abundant", "deficient"][
        _cmp(sum(x for x in range(1, number - 1) if number % x == 0), number)
    ]


def _cmp(x, y):
    return (x > y) - (y > x)

File: haskell/perfect-numbers/src/PerfectNumbers.hs
module PerfectNumbers (classify, Classification (..)) where

data Classification = Deficient | Perfect | Abundant deriving (Enum, Eq, Show)

classify :: Int -> Maybe Classification
classify x = toEnum . fromEnum . compare (sum properDivisors) <$> if x <= 0 then Nothing else Just x
  where
    properDivisors = [y | y <- [1 .. x - 2], x `mod` y == 0]
